manager/src/main/java/org/openremote/manager/alarm/AlarmResourceImpl.java

```diff
@@ -51,79 +55,140 @@ public AlarmResourceImpl(TimerService timerService,
-        } else if (getClientId() != null) {
-            alarm.setSource(CLIENT);
-            alarm.setSourceId(getClientId());
+        try {
+            if (getUserId() != null) {
+                alarm.setSource(MANUAL);
+                alarm.setSourceId(getUserId());
+            } else if (getClientId() != null) {
+                alarm.setSource(CLIENT);
+                alarm.setSourceId(getClientId());
+            }
+
+            if (!isRealmActiveAndAccessible(alarm.getRealm())) {
+                throw new ForbiddenException("Realm '" + alarm.getRealm() + "' is nonexistent, inactive or inaccessible");
+            }
+            return alarmService.sendAlarm(alarm, assetIds);
+        } catch (NullPointerException e) {
+            throw new WebApplicationException(Response.Status.BAD_REQUEST);
+        } catch (ForbiddenException e) {
+            throw new WebApplicationException(Response.Status.FORBIDDEN);
         }
-        return mapExceptions(() -> alarmService.sendAlarm(alarm, assetIds, getUserId()));
     }
 
     @Override
-    public void updateAlarm(RequestParams requestParams, Long alarmId, SentAlarm alarm) {
-        mapExceptions(() -> alarmService.updateAlarm(alarmId, getUserId(), alarm));
+    public void updateAlarm(RequestParams requestParams, Long alarmId, SentAlarm newAlarm) {
+        try {
+            SentAlarm oldAlarm = alarmService.getAlarm(alarmId);
+            if (!isRealmActiveAndAccessible(oldAlarm.getRealm())) {
+                throw new ForbiddenException("Realm '" + oldAlarm.getRealm() + "' is nonexistent, inactive or inaccessible");
+            }
+            alarmService.updateAlarm(oldAlarm, newAlarm);
+        } catch (EntityNotFoundException e) {
+            throw new WebApplicationException(Response.Status.NOT_FOUND);
+        } catch (ForbiddenException e) {
+            throw new WebApplicationException(Response.Status.FORBIDDEN);
+        }
     }
 
     @Override
     public List<AlarmAssetLink> getAssetLinks(RequestParams requestParams, Long alarmId, String realm) {
-        return mapExceptions(() -> alarmService.getAssetLinks(alarmId, getUserId(), realm));
+        try {
+            if (!isRealmActiveAndAccessible(realm)) {
+                throw new ForbiddenException("Realm '" + realm + "' is nonexistent, inactive or inaccessible");
+            }
+            return alarmService.getAssetLinks(alarmId, realm);
+        } catch (EntityNotFoundException e) {
+            throw new WebApplicationException(Response.Status.NOT_FOUND);
+        } catch (ForbiddenException e) {
+            throw new WebApplicationException(Response.Status.FORBIDDEN);
+        }
     }
 
     @Override
     public void setAssetLinks(RequestParams requestParams, List<AlarmAssetLink> links) {
-        mapExceptions(() -> alarmService.linkAssets(links, getUserId()));
+        try {
+            if (links == null || links.isEmpty()) {
+                throw new IllegalArgumentException("Missing links");
```

Comment:
Would make this exception a bit more clear; "No AlarmAssetLink objects were provided."